introduce RequestValidationError and ResponseValidationError exceptions

gweis opened the pull request:
remove add_validation_error_view config directive which can be replaced by standard pyramid custom exception views
re-raise any exceptions thrown in view (to keep view semantics of returning vs throwing HTTPExceptions)
updated tests

gweis commented:
Seems like I overlooked that circle-ci is running the examples. I'll update

zupo commented on pyramid_openapi3/tests/test_validation.py:
```diff
@@ -0,0 +1,198 @@
+        # not enough of pyramid has been set up so we need to render the
+        # exception response ourselves.
+        response.prepare({"HTTP_ACCEPT": "application/json"})
+        self.assertIn("failed", response.json["message"])
```
`failed` is a too generic word. It could be raised by some other error, and this test would think everything is good. Please use a string that is less likely to be present in arbitrary error messages, for example `foo` or `foobar` or `this is a bad request`.

gweis replied:
Yes, I was thinking the same when I wrote that test :). Will update the pull request here.

gweis replied:
fixed

gweis commented:
I have updated the pull request and changed the behaviour a fair bit.
The idea is, that request validation happens just like before, just that `RequestValidationError` which is a subclass of `HTTPException` get's thrown. A side effect of this is, that standard pyramid exception handling can easily handle it, but pretty much any operation in `openapi.yaml` now needs to define a 400 response code to correctly validate a possible an invalid request response.

I have moved the response validation code into a tween, so that other tweens like pyramid_exception_tween to handle the exception and generate a response before response validation kicks in.

There are probably still a few rough edges, but now it covers pretty much all my use cases.

zupo commented:
Great, I'll have a look!

> but pretty much any operation in openapi.yaml now needs to define a 400 response code to correctly validate a possible an invalid request response.

This is fine, we're planning to provide a way to [enforce common responses](https://github.com/Pylons/pyramid_openapi3/issues/#22) soon-ish.

zupo commented on pyramid_openapi3/tween.py:
```diff
@@ -0,0 +1,53 @@
+    The only problem here is, that when response validation fails, we have
+    to return some exception response, with an unknown content type.
+    The advantage is, that this are server errors, and if 500 errors are
+    only possibly due to response validation errors we don't need to document
```
s/possibly/possible

gweis replied:
good catch ... probably also s/this/these ... I'll update

zupo commented:
No changes required on pyramid-realworld-example-app: https://travis-ci.com/niteoweb/pyramid-realworld-example-app/builds/115003852

gweis commented:
> No changes required on pyramid-realworld-example-app

That's great. The only changes that may be needed are replacing `config.add_validation_error_view` with `config.add_exception_view` and maybe adding '400' responses to the projects openapi.yaml.

In a future pull/feature request, I'd like to make the request/response validation optional. E.g.: by configuring the view with a `openapi=(True, False) # enable request, disable response validation`. Not sure yet, but I may have some rare use cases for that, and it would be a very minimal change to support that.

zupo commented:
> by configuring the view with a openapi=(True, False)

As long as `openapi` stays enough for the default behaviour (no need to say `(True, True)`, then +1

zupo commented:
Still need to test this against our private codebase that is currently the biggest user of pyramid_openapi3.

gweis commented:
> Still need to test this against our private codebase that is currently the biggest user of pyramid_openapi3

Yes, that is the intention. No backwards incompatible changes necessary. (and barely a performance hit, as this would be parsed on startup).

MatthewWilkes commented on pyramid_openapi3/tests/test_validation.py:
```diff
@@ -0,0 +1,255 @@
+        from pyramid.httpexceptions import HTTPBadRequest
+
+        self._add_view(
+            lambda *arg: (_ for _ in ()).throw(HTTPBadRequest("bad foo request"))
```
Is the generator `throw` here meaningful, or is it a workaround for not being able to `raise` inside a `lambda`? I would think that a normal function would be a clearer analogue to a normal view.

gweis replied:
Agree, took me a while myself to understand this `generator throw` hack. 
Wasn't sure what to do. Either use this construct to raise an exception in a `lambda` or break the pattern used in other tests and use a local function. (updated pull request)

MatthewWilkes commented:
I've gone over the changes here and it looks really good to me. My only concern was a minor readability item in the tests.

dz0ny commented on pyramid_openapi3/__init__.py:
```diff
@@ -42,48 +39,24 @@ def openapi_view(view: View, info: ViewDeriverInfo) -> t.Optional[View]:
-
-            return response
+            if request.openapi_validated.errors:
+                raise RequestValidationError(errors=request.openapi_validated.errors)
```
If you down the line process exception in Sentry, Stackdriver etc, it's really helpful if you also have request.openapi_validated.body and similar, so I would pass request.openapi_validated and then defer parsing values in the exception handler.

gweis replied:
As far as I know, in case of a RequestValidationError, `request.openapi_validated.parameters` will only ever have valid parameters and `request.openapi_validated.body` is only populated when the `body` is valid. So any parameter or body content that errors won't be available in on the `ValidationResult`. So all you would get are valid parameters and valid body. 

I was not sure, whether to pass the `ValidationResult` or just the errors to the exception. I couldn't find any useful values on a `ValidationResult` other than errors (everything I would probably interrogate would be available on the request object only), so I opted for that to be on the safe side. My concern was rather about, what if it is a really large body object, and do I really want that to be passed around? (probably not an issue, but I assumed not passing it through is safer :) )

dz0ny replied:
> My concern was rather about, what if it is a really large body object, and do I really want that to be passed around?
It is not passed around, it's reference is passed around.

> I couldn't find any useful values on a ValidationResult other than errors.
Then it's fine I assumed custom formattes might be able to pass things into the object, but that's not the case.

dz0ny commented on pyramid_openapi3/exceptions.py:
```diff
@@ -0,0 +1,39 @@
+    def __init__(self, *args, errors, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.errors = errors
+        self.detail = self.message = "\n".join(str(e) for e in errors)
```
should be deferred to handler

gweis replied:
You mean the exception view handler? The pyramid default exception view handler does not handle exceptions with an `error` attribute. Are you suggesting to add a default exception handler to this library?

dz0ny replied:
No, the exact line ```self.detail = self.message = "\n".join(str(e) for e in errors)``` you are creating object in memory that are not even needed yet.

gweis replied:
I was trying to keep the pyramid httpexception contract up here with some useful information from validation errors, and not re-implementing all the templating that's implemented in pyramid http exception/response. 

I'll have a think about whether it's possible to delay that `detail` / `message` generation. The only way I see for now is to use a `@property` wrapper around `.message` and `.detail` . Maybe there are better ways?

dz0ny commented on examples/todoapp/app.py:
```diff
@@ -49,12 +51,12 @@ def post(request: Request) -> t.List[t.Dict["str", "str"]]:
+@exception_view_config(RequestValidationError, renderer="json")
 def openapi_validation_error(
-    context: DefaultRootFactory, request: Request
+    context: HTTPException, request: Request
```
context: https://docs.pylonsproject.org/projects/pyramid/en/latest/glossary.html#term-context is a bit special word used in Pyramid, **exception**  is totally fine

gweis replied:
Didn't want to change original code here.

dz0ny replied:
I, see it's event implied in https://github.com/Pylons/pyramid/blob/master/src/pyramid/httpexceptions.py#L1326 that it can be both.

As it's not consistent anyway we can ignore it.

dz0ny commented on pyramid_openapi3/tween.py:
```diff
@@ -0,0 +1,53 @@
+        # If there is no exception view, we also see request validation errors here
+        except ResponseValidationError:
+            exc_info = sys.exc_info()
+            try:
```
You can reuse [_error_handler](https://docs.pylonsproject.org/projects/pyramid/en/latest/_modules/pyramid/tweens.html?highlight=_error_handler)

gweis replied:
Not sure whether it is good practice to re-use `private` functions?

dz0ny replied:
I've run it on an internal project and it seems it breaks badly because exception handling is reimplemented. Will have more info in a couple of days.

gweis replied:
Maybe there is somewhere a custom exception view handler that returns the `ResponseValidationResult` instead of a Response / HTTPException instance? Otherwise I can't see why there would be a `ResponseValidationResult` returned by the debug_toolbar tween handler.

zupo commented:
@gweis: quick update: got the private project's build green, now need to get it through review and pushed to production.

dz0ny commented on pyramid_openapi3/tween.py:
```diff
@@ -0,0 +1,53 @@
+    def excview_tween(request) -> Response:
+        try:
+            response = handler(request)
+            if not request.environ.get("pyramid_openapi3.validate_response"):
```
**pyramid_openapi3.validate_response** infers that only response will not be validated, but the code actually ignores the request too.

gweis replied:
That's a bit of a source for confusion here... the Request is validated in the view wrapper, but the response is validated after the exception view handler ran (if there is any)

dz0ny commented:
@gweis Can you explain a bit what were your motivations for re-implementing error handler and not having just standard tween? I don't understand why is this better?

gweis commented:
@dzony The original problem was, that within pyramid apps where not every view is an openapi3 view, You need some way of deciding when to validate the request. That's done in this pkg via the view wrapper.
However I then noticed, that when I throw exceptions in my openapi enabled view these are caught by the same wrapper and returned as response, meaning that I would have now two ways to handle exceptions. One for normal views, where a raised Exception "bubbles" up into the pyramid exception view handler, and raised Exceptions in an openapi view, where any exception I raise, get's returned and does not invoke the exception view handler.
So I was looking for a way to keep usual pyramid behaviour with raised or returned HTTPExceptions, use pyramids exception view handling mechanisms to handle errors, and still validate request / response / properly. This meant it was necessary to move the response validation after error view handler have been processed.
In short the aim is to align as much as possible with documented behaviour of the pyramid request chain, and re-using mechanism that already exist in pyramid instead of changing the request/response processing (esp. view error handling, regarding raised/returned exceptions).

mmerickel commented on pyramid_openapi3/tween.py:
```diff
@@ -0,0 +1,53 @@
+            if result.errors:
+                raise ResponseValidationError(errors=result.errors)
+        # If there is no exception view, we also see request validation errors here
+        except ResponseValidationError:
```
I think a test is missing here for `if not request.environ.get('pyramid_openapi3.validate_response')` incase the handler raises one of these errors for some reason on a view that's not protected by the spec.

gweis replied:
Not quite sure whether I follow here, but I have added another test. https://github.com/Pylons/pyramid_openapi3/pull/26/commits/48b804907c5dfb430f77c8d936f043e6a5362437

Otherwise I think a non openapi view should not raise a `ResponseValidationError`. Maybe we should think about that in case there is a real use case for it, otherwise I'd consider throwing random `ResponseValidationError` instances in user code as a bug.

mmerickel commented on pyramid_openapi3/tween.py:
```diff
@@ -0,0 +1,53 @@
+            try:
+                response = request.invoke_exception_view(exc_info)
+            except HTTPNotFound:
+                reraise(*exc_info)
```
a `finally` is missing here to `del exc_info` to avoid cycles on exc_info stack frames referencing the current stack frame... I will say I'm sensitive to this because of python 2 and I'm not positive it's an issue in python 3 - but I wanted to bring it up to make sure someone double checks :-)

gweis replied:
I don't think a `del exc_info` is needed here, but I am not really sure, so I have added it anyway (can't hurt ).

gweis commented:
Had to update version pins for Pipfile, otherwise it would not build on travis.

zupo commented:
@gweis: I finally found enough time to give this PR another rundown and test it in our private codebase. It seems to work exactly as advertised , so I merged it. I'm terribly sorry it took me this long! Thanks a million for your contribution, the package is so much better off with this change!

zupo commented:
I've cut a new release that includes this PR: https://pypi.org/project/pyramid-openapi3/0.4.0/
